In [1]:
import pandas as pd

In [82]:
df1=pd.read_csv("/content/drive/MyDrive/NIR/data1.csv")

# 1) How to identify the Records (Order ID + Product ID combination) present in data1 but missing in data2?

In [83]:
df1

Order ID       Product ID  Qty
0     CA-2014-100006  TEC-PH-10002075    3
1     CA-2014-100090  FUR-TA-10003715    3
2     CA-2014-100090  OFF-BI-10001597    6
3     CA-2014-100293  OFF-PA-10000176    6
4     CA-2014-100328  OFF-BI-10000343    1
...              ...              ...  ...
9473  US-2017-169551  OFF-PA-10004100    3
9474  US-2017-169551  OFF-ST-10004835    3
9475  US-2017-169551  TEC-AC-10002018    3
9476  US-2017-169551  TEC-AC-10003033    2
9477  US-2017-169551  TEC-PH-10001363    2

[9478 rows x 3 columns]

In [84]:
df2=pd.read_csv("/content/drive/MyDrive/NIR/data2.csv")
df2

Order ID       Product ID  Qty
0     CA-2014-100090  FUR-TA-10003715    3
1     CA-2014-100090  OFF-BI-10001597    6
2     CA-2014-100293  OFF-PA-10000176    6
3     CA-2014-100328  OFF-BI-10000343    1
4     CA-2014-100363  OFF-FA-10000611    2
...              ...              ...  ...
9474  US-2017-169551  OFF-PA-10004100    3
9475  US-2017-169551  OFF-ST-10004835    3
9476  US-2017-169551  TEC-AC-10002018    3
9477  US-2017-169551  TEC-AC-10003033    2
9478  US-2017-169551  TEC-PH-10001363    2

[9479 rows x 3 columns]

In [85]:
df1["Record"] = df1["Order ID"]+df1["Product ID"]
df2["Record"] = df2["Order ID"]+df2["Product ID"]

In [8]:
df_result = df1.merge(df2.drop_duplicates(), on=['Order ID','Product ID'], how='left', indicator=True)

In [14]:
df_result.head(50)

Order ID       Product ID  Qty_x                       Record_x  \
0   CA-2014-100006  TEC-PH-10002075      3  CA-2014-100006TEC-PH-10002075   
1   CA-2014-100090  FUR-TA-10003715      3  CA-2014-100090FUR-TA-10003715   
2   CA-2014-100090  OFF-BI-10001597      6  CA-2014-100090OFF-BI-10001597   
3   CA-2014-100293  OFF-PA-10000176      6  CA-2014-100293OFF-PA-10000176   
4   CA-2014-100328  OFF-BI-10000343      1  CA-2014-100328OFF-BI-10000343   
5   CA-2014-100363  OFF-FA-10000611      2  CA-2014-100363OFF-FA-10000611   
6   CA-2014-100363  OFF-PA-10004733      3  CA-2014-100363OFF-PA-10004733   
7   CA-2014-100391  OFF-PA-10001471      2  CA-2014-100391OFF-PA-10001471   
8   CA-2014-100678  FUR-CH-10002602      3  CA-2014-100678FUR-CH-10002602   
9   CA-2014-100678  OFF-AR-10001868      2  CA-2014-100678OFF-AR-10001868   
10  CA-2014-100678  OFF-EN-10000056      3  CA-2014-100678OFF-EN-10000056   
11  CA-2014-100678  TEC-AC-10000474      3  CA-2014-100678TEC-AC-10000474   
12  CA-2014-100706  FUR-FU-10002268      6  CA-2014-100706FUR-FU-10002268   
13  CA-2014-100762  OFF-AR-10000380      4  CA-2014-100762OFF-AR-10000380   
14  CA-2014-100762  OFF-LA-10003930      2  CA-2014-100762OFF-LA-10003930   
15  CA-2014-100762  OFF-PA-10004082      2  CA-2014-100762OFF-PA-10004082   
16  CA-2014-100860  OFF-LA-10001982      5  CA-2014-100860OFF-LA-10001982   
17  CA-2014-100867  TEC-PH-10004922      6  CA-2014-100867TEC-PH-10004922   
18  CA-2014-100881  TEC-PH-10003273      3  CA-2014-100881TEC-PH-10003273   
19  CA-2014-100895  OFF-AR-10004511      2  CA-2014-100895OFF-AR-10004511   
20  CA-2014-100895  OFF-ST-10001490      2  CA-2014-100895OFF-ST-10001490   
21  CA-2014-100895  TEC-PH-10001425      3  CA-2014-100895TEC-PH-10001425   
22  CA-2014-100916  FUR-TA-10004607      4  CA-2014-100916FUR-TA-10004607   
23  CA-2014-100916  OFF-AR-10001022      1  CA-2014-100916OFF-AR-10001022   
24  CA-2014-100916  OFF-ST-10003479      5  CA-2014-100916OFF-ST-10003479   
25  CA-2014-100972  OFF-PA-10000357      3  CA-2014-100972OFF-PA-10000357   
26  CA-2014-101147  OFF-AP-10004249      1  CA-2014-101147OFF-AP-10004249   
27  CA-2014-101175  OFF-ST-10004950      6  CA-2014-101175OFF-ST-10004950   
28  CA-2014-101266  OFF-PA-10002986      2  CA-2014-101266OFF-PA-10002986   
29  CA-2014-101364  OFF-BI-10003984     13  CA-2014-101364OFF-BI-10003984   
30  CA-2014-101392  OFF-ST-10000532      7  CA-2014-101392OFF-ST-10000532   
31  CA-2014-101462  FUR-FU-10000409      4  CA-2014-101462FUR-FU-10000409   
32  CA-2014-101476  TEC-MA-10000029      1  CA-2014-101476TEC-MA-10000029   
33  CA-2014-101560  FUR-FU-10003773      5  CA-2014-101560FUR-FU-10003773   
34  CA-2014-101560  OFF-AR-10004757      4  CA-2014-101560OFF-AR-10004757   
35  CA-2014-101560  OFF-BI-10000309      3  CA-2014-101560OFF-BI-10000309   
36  CA-2014-101560  OFF-PA-10001363      7  CA-2014-101560OFF-PA-10001363   
37  CA-2014-101602  FUR-CH-10004675      5  CA-2014-101602FUR-CH-10004675   
38  CA-2014-101602  TEC-PH-10000169      3  CA-2014-101602TEC-PH-10000169   
39  CA-2014-101770  OFF-BI-10001097      1  CA-2014-101770OFF-BI-10001097   
40  CA-2014-101833  OFF-PA-10003591      3  CA-2014-101833OFF-PA-10003591   
41  CA-2014-101931  FUR-BO-10001337      6  CA-2014-101931FUR-BO-10001337   
42  CA-2014-101931  OFF-BI-10004140      2  CA-2014-101931OFF-BI-10004140   
43  CA-2014-101931  OFF-ST-10003442      5  CA-2014-101931OFF-ST-10003442   
44  CA-2014-101931  OFF-SU-10000646      2  CA-2014-101931OFF-SU-10000646   
45  CA-2014-101931  OFF-SU-10002301      2  CA-2014-101931OFF-SU-10002301   
46  CA-2014-102008  OFF-PA-10004092      1  CA-2014-102008OFF-PA-10004092   
47  CA-2014-102085  OFF-LA-10002475      4  CA-2014-102085OFF-LA-10002475   
48  CA-2014-102274  OFF-AR-10002833      3  CA-2014-102274OFF-AR-10002833   
49  CA-2014-102274  OFF-PA-10004359      6  CA-2014-102274OFF-PA-10004359   

    Qty_y                       Record_y     _merge  
0     NaN                   

In [13]:
df_result["_merge"].value_counts()

both          8971
left_only      507
right_only       0
Name: _merge, dtype: int64

 So, from the above dataframe df_results,therefore the number of records which are present in data1 but not in data2 is **507**

# 2) How to identify the Records (Order ID + Product ID combination) missing in data2 but present in data1?

In [61]:
df_result1 = df2.merge(df1.drop_duplicates(), on=['Order ID','Product ID'], how='left', indicator=True)

In [63]:
df_result1

Order ID       Product ID  Qty_x                       Record_x  \
0     CA-2014-100090  FUR-TA-10003715      3  CA-2014-100090FUR-TA-10003715   
1     CA-2014-100090  OFF-BI-10001597      6  CA-2014-100090OFF-BI-10001597   
2     CA-2014-100293  OFF-PA-10000176      6  CA-2014-100293OFF-PA-10000176   
3     CA-2014-100328  OFF-BI-10000343      1  CA-2014-100328OFF-BI-10000343   
4     CA-2014-100363  OFF-FA-10000611      2  CA-2014-100363OFF-FA-10000611   
...              ...              ...    ...                            ...   
9474  US-2017-169551  OFF-PA-10004100      3  US-2017-169551OFF-PA-10004100   
9475  US-2017-169551  OFF-ST-10004835      3  US-2017-169551OFF-ST-10004835   
9476  US-2017-169551  TEC-AC-10002018      3  US-2017-169551TEC-AC-10002018   
9477  US-2017-169551  TEC-AC-10003033      2  US-2017-169551TEC-AC-10003033   
9478  US-2017-169551  TEC-PH-10001363      2  US-2017-169551TEC-PH-10001363   

      Qty_y                       Record_y _merge  
0       3.0  CA-2014-100090FUR-TA-10003715   both  
1       6.0  CA-2014-100090OFF-BI-10001597   both  
2       6.0  CA-2014-100293OFF-PA-10000176   both  
3       1.0  CA-2014-100328OFF-BI-10000343   both  
4       2.0  CA-2014-100363OFF-FA-10000611   both  
...     ...                            ...    ...  
9474    3.0  US-2017-169551OFF-PA-10004100   both  
9475    3.0  US-2017-169551OFF-ST-10004835   both  
9476    3.0  US-2017-169551TEC-AC-10002018   both  
9477    2.0  US-2017-169551TEC-AC-10003033   both  
9478    2.0  US-2017-169551TEC-PH-10001363   both  

[9479 rows x 7 columns]

In [62]:
df_result1["_merge"].value_counts()

both          8971
left_only      508
right_only       0
Name: _merge, dtype: int64

So, from the above dataframe df_results,therefore the number of records which are present in data2 but not in data1 is **508**

# 3) Find the Sum of the total Qty of Records missing in data1 but present in data2?


In [64]:
df_result1

Order ID       Product ID  Qty_x                       Record_x  \
0     CA-2014-100090  FUR-TA-10003715      3  CA-2014-100090FUR-TA-10003715   
1     CA-2014-100090  OFF-BI-10001597      6  CA-2014-100090OFF-BI-10001597   
2     CA-2014-100293  OFF-PA-10000176      6  CA-2014-100293OFF-PA-10000176   
3     CA-2014-100328  OFF-BI-10000343      1  CA-2014-100328OFF-BI-10000343   
4     CA-2014-100363  OFF-FA-10000611      2  CA-2014-100363OFF-FA-10000611   
...              ...              ...    ...                            ...   
9474  US-2017-169551  OFF-PA-10004100      3  US-2017-169551OFF-PA-10004100   
9475  US-2017-169551  OFF-ST-10004835      3  US-2017-169551OFF-ST-10004835   
9476  US-2017-169551  TEC-AC-10002018      3  US-2017-169551TEC-AC-10002018   
9477  US-2017-169551  TEC-AC-10003033      2  US-2017-169551TEC-AC-10003033   
9478  US-2017-169551  TEC-PH-10001363      2  US-2017-169551TEC-PH-10001363   

      Qty_y                       Record_y _merge  
0       3.0  CA-2014-100090FUR-TA-10003715   both  
1       6.0  CA-2014-100090OFF-BI-10001597   both  
2       6.0  CA-2014-100293OFF-PA-10000176   both  
3       1.0  CA-2014-100328OFF-BI-10000343   both  
4       2.0  CA-2014-100363OFF-FA-10000611   both  
...     ...                            ...    ...  
9474    3.0  US-2017-169551OFF-PA-10004100   both  
9475    3.0  US-2017-169551OFF-ST-10004835   both  
9476    3.0  US-2017-169551TEC-AC-10002018   both  
9477    2.0  US-2017-169551TEC-AC-10003033   both  
9478    2.0  US-2017-169551TEC-PH-10001363   both  

[9479 rows x 7 columns]

In [72]:
df_new=df_result1.loc[df_result1["_merge"]=="left_only"]

In [73]:
df_new

Order ID       Product ID  Qty_x                       Record_x  \
11    CA-2014-100706  TEC-AC-10001314      2  CA-2014-100706TEC-AC-10001314   
14    CA-2014-100762  OFF-PA-10001815      3  CA-2014-100762OFF-PA-10001815   
30    CA-2014-101427  OFF-AR-10002257      3  CA-2014-101427OFF-AR-10002257   
56    CA-2014-102652  FUR-FU-10001918      7  CA-2014-102652FUR-FU-10001918   
63    CA-2014-102869  OFF-PA-10000788      3  CA-2014-102869OFF-PA-10000788   
...              ...              ...    ...                            ...   
9428  US-2017-165344  OFF-BI-10003196     10  US-2017-165344OFF-BI-10003196   
9433  US-2017-165358  TEC-CO-10001943      5  US-2017-165358TEC-CO-10001943   
9455  US-2017-167920  OFF-AP-10000159      5  US-2017-167920OFF-AP-10000159   
9471  US-2017-169502  OFF-AP-10001947      5  US-2017-169502OFF-AP-10001947   
9473  US-2017-169551  FUR-BO-10001519      3  US-2017-169551FUR-BO-10001519   

      Qty_y Record_y     _merge  
11      NaN      NaN  left_only  
14      NaN      NaN  left_only  
30      NaN      NaN  left_only  
56      NaN      NaN  left_only  
63      NaN      NaN  left_only  
...     ...      ...        ...  
9428    NaN      NaN  left_only  
9433    NaN      NaN  left_only  
9455    NaN      NaN  left_only  
9471    NaN      NaN  left_only  
9473    NaN      NaN  left_only  

[508 rows x 7 columns]

In [74]:
df_new["Qty_x"].sum()

1956

Sum of the total Qty of Records missing in data1 but present in data2 is **1956**


# 4) Find the total number of unique records (Order ID + Product ID combination) present in the combined dataset of data1 and data2?

In [75]:
df_result2 = df2.merge(df1.drop_duplicates(), on=['Order ID','Product ID'], how='inner', indicator=True)

In [76]:
df_result2

Order ID       Product ID  Qty_x                       Record_x  \
0     CA-2014-100090  FUR-TA-10003715      3  CA-2014-100090FUR-TA-10003715   
1     CA-2014-100090  OFF-BI-10001597      6  CA-2014-100090OFF-BI-10001597   
2     CA-2014-100293  OFF-PA-10000176      6  CA-2014-100293OFF-PA-10000176   
3     CA-2014-100328  OFF-BI-10000343      1  CA-2014-100328OFF-BI-10000343   
4     CA-2014-100363  OFF-FA-10000611      2  CA-2014-100363OFF-FA-10000611   
...              ...              ...    ...                            ...   
8966  US-2017-169551  OFF-PA-10004100      3  US-2017-169551OFF-PA-10004100   
8967  US-2017-169551  OFF-ST-10004835      3  US-2017-169551OFF-ST-10004835   
8968  US-2017-169551  TEC-AC-10002018      3  US-2017-169551TEC-AC-10002018   
8969  US-2017-169551  TEC-AC-10003033      2  US-2017-169551TEC-AC-10003033   
8970  US-2017-169551  TEC-PH-10001363      2  US-2017-169551TEC-PH-10001363   

      Qty_y                       Record_y _merge  
0         3  CA-2014-100090FUR-TA-10003715   both  
1         6  CA-2014-100090OFF-BI-10001597   both  
2         6  CA-2014-100293OFF-PA-10000176   both  
3         1  CA-2014-100328OFF-BI-10000343   both  
4         2  CA-2014-100363OFF-FA-10000611   both  
...     ...                            ...    ...  
8966      3  US-2017-169551OFF-PA-10004100   both  
8967      3  US-2017-169551OFF-ST-10004835   both  
8968      3  US-2017-169551TEC-AC-10002018   both  
8969      2  US-2017-169551TEC-AC-10003033   both  
8970      2  US-2017-169551TEC-PH-10001363   both  

[8971 rows x 7 columns]

In [77]:
df_result2["_merge"].value_counts()

both          8971
left_only        0
right_only       0
Name: _merge, dtype: int64

total number of unique records (Order ID + Product ID combination) present in the combined dataset of data1 and data2 is **8971**